src/dag.c

```diff
@@ -901,399 +901,447 @@ void RedisAI_Disconnected(RedisModuleCtx *ctx, RedisModuleBlockedClient *bc) {
+			int *instance = AI_dictGetVal(mangled_entry);
+			RedisModuleString *mangled_key = RedisModule_CreateStringPrintf(ctx, "%s%04d", key, *instance);
+			const char* mangled_key_str = RedisModule_StringPtrLen(mangled_key, NULL);
+			AI_dictAdd(mangled_persisted, (void *)mangled_key_str, (void *)1);
+			entry = AI_dictNext(iter);
+		}
+		AI_dictReleaseIterator(iter);
+	}
+
+	AI_dictRelease(rinfo->dagTensorsPersistedContext);
+	rinfo->dagTensorsPersistedContext = mangled_persisted;
+
+	{
+		AI_dictIterator *iter = AI_dictGetSafeIterator(mangled_tensors);
+		AI_dictEntry *entry = AI_dictNext(iter);
+		while (entry) {
+			int *val = (int *)AI_dictGetVal(entry);
+			RedisModule_Free(val);
+			entry = AI_dictNext(iter);
+		}
+		AI_dictReleaseIterator(iter);
+	}
+	AI_dictRelease(mangled_tensors);
+	mangled_tensors = NULL;
+
+	for (long long i=0; i<array_len(rinfo->dagOps); i++) {
+		if (rinfo->dagOps[i]->devicestr == NULL) {
+			rinfo->dagOps[i]->devicestr = "CPU";
+		}
+	}
+	return REDISMODULE_OK;
+}
 
-  for (long long i=1; i<ndevices; i++) {
-    RedisAI_RunInfo *rinfo_copy;
-    RAI_ShallowCopyDagRunInfo(&rinfo_copy, rinfo);
-    rinfo_copies = array_append(rinfo_copies, rinfo_copy);
-  }
- 
-  for (long long i=0; i<ndevices; i++) {
-    RedisAI_RunInfo *rinfo = rinfo_copies[i];
-    for (long long j=0; j<rinfo->dagOpCount; j++) {
-      if (strcasecmp(rinfo->dagOps[j]->devicestr, devices[i]) == 0) {
-        rinfo->dagDeviceOps = array_append(rinfo->dagDeviceOps, rinfo->dagOps[j]);
-      }
-    }
+static int DAG_InsertDAGToQueue (RedisAI_RunInfo *rinfo) {
+	const char **devices = array_new(const char *, 10);
+
+	for (long long i=0; i<array_len(rinfo->dagOps); i++) {
+		const char* devicestr = rinfo->dagOps[i]->devicestr;
+		bool found = false;
+		for (long long j=0; j<array_len(devices); j++) {
+			if (strcasecmp(devicestr, devices[j]) == 0) {
+				found = true;
+				break;
+			}
+		}
+		if (!found) {
+			devices = array_append(devices, devicestr);
+		}
+	}
+
+	size_t ndevices = array_len(devices);
+
+	*rinfo->dagRefCount = ndevices;
+
+	RedisAI_RunInfo **rinfo_copies = array_new(RedisAI_RunInfo*, ndevices);
+	rinfo_copies = array_append(rinfo_copies, rinfo);
+
+	for (long long i=1; i<ndevices; i++) {
+		RedisAI_RunInfo *rinfo_copy;
+		RAI_ShallowCopyDagRunInfo(&rinfo_copy, rinfo);
+		rinfo_copies = array_append(rinfo_copies, rinfo_copy);
+	}
+
+	for (long long i=0; i<ndevices; i++) {
+		RedisAI_RunInfo *rinfo = rinfo_copies[i];
+		for (long long j=0; j<rinfo->dagOpCount; j++) {
+			if (strcasecmp(rinfo->dagOps[j]->devicestr, devices[i]) == 0) {
+				rinfo->dagDeviceOps = array_append(rinfo->dagDeviceOps, rinfo->dagOps[j]);
+			}
+		}
+
+		rinfo->dagDeviceOpCount = array_len(rinfo->dagDeviceOps);
+	}
+
+	for (long long i=0; i<ndevices; i++) {
+		const char* devicestr = devices[i];
+		RunQueueInfo *run_queue_info = NULL;
+		if (ensureRunQueue(devicestr, &run_queue_info) == REDISMODULE_ERR) {
+			RAI_FreeRunInfo(rinfo);
+			return REDISMODULE_ERR;
```

Comment:
first validate queues, enqueue after.
In case of failure set dag global error and call the on finish